chainer/optimizer_hooks/gradient_clipping.py

```diff
@@ -52,19 +82,21 @@ def __init__(self, threshold):
         self.threshold = threshold
 
     def __call__(self, opt):
-        sqnorm = _sum_sqnorm([p.grad for p in opt.target.params(False)])
-        device = backend.get_device_from_array(sqnorm)
+        sqnorm, device = _sum_sqnorm_grads(list(opt.target.params(False)))
```

Comment:
>Why not pass the parameters as a generator instead of creating a temporary list?

`assert(len(params))` wouldn't work. A list is easier to manipulate than a generator.
 
>A minor but how about returning the norm from this helper instead of it squared?

That would require an extra `using_device`.

Reply:
I think that performance matters more for the users of this hook. But also, I don't think that this assertion must necessarily hold for the hook to be well behaived? This hook could work "just fine" even when there are no parameters to the target. The norm would be 0 but this may also happen when `target.params` is not empty. PyTorch seem to add an eps to the norm during the division below to get around this.

Reply:
> That would require an extra using_device.

I see, it makes sense.

Reply:
I see, maybe I should remove the assertion.
But does `list()` affect performance much?
Lists are less bug-prone, which I think is more important than saving ~100ns per call.

Reply:
Removed the assertion, and added TODOs.

Reply:
I'm not promoting premature optimizations but it just seemed quite obvious to me in this situation where we're just once iterating over it. I did not see the risk of bugs however. Would you mind explaining in what sense this is bug-prone?

Reply:
You could edit `_sum_sqnorm_grads` without understanding `params` was actually a generator.

Reply:
Ah, okay, but this is an internal function that's only mean to be called from a single location (it could, possibly, be inlined). Isn't a comment enough? We could also do a `isinstance(params, types.GeneratoType)` but that's probably overkill...

Reply:
Honestly I don't understand why we need this long discussion on this point (i.e. just 100ns per call). Do I possibly underestimate the effect?

Reply:
Me neither, I just don't see why, when a comment should do to prevent errors. But okay let us leave is as if.